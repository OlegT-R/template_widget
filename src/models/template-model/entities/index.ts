export * from './node';
export * from './template';
export * from './textarea';
export * from './variables';
