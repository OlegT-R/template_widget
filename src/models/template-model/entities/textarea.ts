export interface ITextArea {
    id: string;
    value: string;
}
