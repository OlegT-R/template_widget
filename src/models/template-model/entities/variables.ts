
export type VariableNamesType = string[];

export interface IVariables {
  [key: string]: string,
}
