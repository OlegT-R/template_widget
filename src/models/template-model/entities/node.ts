import { ITextArea } from './textarea';

export interface IIfElse {
  id: string,
  if: NodeType;
  then: NodeType;
  else: NodeType;
}

export type NodeType = Array<ITextArea | IIfElse>;
