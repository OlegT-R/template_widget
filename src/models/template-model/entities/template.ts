import { IIfElse, NodeType } from './node';
import { ITextArea } from './textarea';

export interface ITemplate {
  block: NodeType;
}

export type NodeItem = ITextArea | IIfElse;

export function isTextArea(node: NodeItem): node is ITextArea {
  return (node as ITextArea).value !== undefined;
}

export function isIfElse(node: NodeItem): node is IIfElse {
  return (node as IIfElse).if !== undefined;
}
