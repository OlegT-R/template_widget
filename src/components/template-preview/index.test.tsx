import React from 'react';

import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { TemplatePreview } from '.';
import { ITemplate } from '../../models/template-model/entities';

const variables = ['var1', 'var2', 'var12'];

const template: ITemplate = {
  block: [
    {
      id: 'Block1Start',
      value: 'Var1{var1}',
    },
    {
      id: 'IfElse1',
      if: [
        {
          id: 'Block2If',
          value: '{var2}',
        },
      ],
      then: [
        {
          id: 'block2',
          value: 'then{var2}',
        },
      ],
      else: [
        {
          id: 'Block2Else',
          value: 'else2',
        },
      ],
    },
    {
      id: 'Block1End',
      value: 'Var12{var12}',
    },
  ],
};

describe('<TemplatePreview />', () => {
  test('Component structure: Should contain textarea and 3 inputs', () => {
    const component: any = mount(
      <TemplatePreview template={template} arrVarNames={variables} />
    );
    // check existing 1 textarea
    const textarea = component.find('textarea');
    expect(textarea.length).toBe(1);

    // check existing 3 input -> 3 vars
    const inputs = component.find('input');
    expect(inputs.length).toBe(3);
    const inputGroup = component.find('ant-input-group');

    // check inputs to expecting var names
    inputGroup.map((group: any, index: number) => {
      const inputLabel = group.find('.ant-input-group-addon').text();
      expect(inputLabel).toEqual(variables[index]);
    });
    expect(toJson(component)).toMatchSnapshot();
  });

  test('Should update message after type', () => {
    const component: any = mount(
      <TemplatePreview template={template} arrVarNames={variables} />
    );
    const { value } = component.find('textarea').props();
    expect(value).toEqual('Var1else2Var12');

    // set var2 variable to UPDATE, expecting then thread
    const input = component
      .find('input')
      .at(1)
      .simulate('change', { target: { value: 'UPDATE' } });
    component.update();
    const { value: updatedValue } = component.find('textarea').props();
    expect(updatedValue).toEqual('Var1thenUPDATEVar12');
  });
});
