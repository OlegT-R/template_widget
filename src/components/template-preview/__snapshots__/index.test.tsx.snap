// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplatePreview /> Component structure: Should contain textarea and 3 inputs 1`] = `
<TemplatePreview
  arrVarNames={
    Array [
      "var1",
      "var2",
      "var12",
    ]
  }
  template={
    Object {
      "block": Array [
        Object {
          "id": "Block1Start",
          "value": "Var1{var1}",
        },
        Object {
          "else": Array [
            Object {
              "id": "Block2Else",
              "value": "else2",
            },
          ],
          "id": "IfElse1",
          "if": Array [
            Object {
              "id": "Block2If",
              "value": "{var2}",
            },
          ],
          "then": Array [
            Object {
              "id": "block2",
              "value": "then{var2}",
            },
          ],
        },
        Object {
          "id": "Block1End",
          "value": "Var12{var12}",
        },
      ],
    }
  }
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bxivhb",
            "isStatic": true,
            "lastClassName": "dBPXXn",
            "rules": Array [
              "
  width: 100%;
  max-width: 600px;
  height: auto;
  min-height: 200px;
  margin: 0 auto;
  position: relative;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bxivhb",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bxivhb dBPXXn"
      >
        <Styled(TextArea)
          autosize={false}
          rows={5}
          value="Var1else2Var12"
        >
          <StyledComponent
            autosize={false}
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-ifAKCX",
                  "isStatic": true,
                  "lastClassName": "eLaMHS",
                  "rules": Array [
                    "
  resize: none;
",
                  ],
                },
                "displayName": "Styled(TextArea)",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-ifAKCX",
                "target": [Function],
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            rows={5}
            value="Var1else2Var12"
          >
            <TextArea
              autosize={false}
              className="sc-ifAKCX eLaMHS"
              rows={5}
              value="Var1else2Var12"
            >
              <ReactResizeObserver
                disabled={true}
                onResize={[Function]}
              >
                <textarea
                  className="ant-input sc-ifAKCX eLaMHS"
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  rows={5}
                  style={Object {}}
                  value="Var1else2Var12"
                />
              </ReactResizeObserver>
            </TextArea>
          </StyledComponent>
        </Styled(TextArea)>
        <VariableInputs
          arrVarNames={
            Array [
              "var1",
              "var2",
              "var12",
            ]
          }
          onChangeVariable={[Function]}
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-htpNat",
                    "isStatic": true,
                    "lastClassName": "ecdxFb",
                    "rules": Array [
                      "
  padding: 10px 0;
  position: relative;
  width: 100%;
  display: flex;
  margin-top: 10px;
  flex-wrap: wrap;
  align-items: center;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-htpNat",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-htpNat ecdxFb"
              >
                <styled.span>
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bwzfXH",
                          "isStatic": true,
                          "lastClassName": "cFLOXe",
                          "rules": Array [
                            "
  padding-right: 5px;
  font-size: 16px;
  font-weight: bold;
",
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bwzfXH",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="sc-bwzfXH cFLOXe"
                    >
                      Variables:
                    </span>
                  </StyledComponent>
                </styled.span>
                <Styled(Input)
                  addonBefore="var1"
                  key="var1"
                  onChange={[Function]}
                >
                  <StyledComponent
                    addonBefore="var1"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "Group": [Function],
                        "Password": [Function],
                        "Search": [Function],
                        "TextArea": [Function],
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bdVaJa",
                          "isStatic": true,
                          "lastClassName": "krKcVG",
                          "rules": Array [
                            "
  margin: 5px 5px 0 0;
  width: 200px;
",
                          ],
                        },
                        "displayName": "Styled(Input)",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bdVaJa",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onChange={[Function]}
                  >
                    <Input
                      addonBefore="var1"
                      className="sc-bdVaJa krKcVG"
                      disabled={false}
                      onChange={[Function]}
                      type="text"
                    >
                      <span
                        className="sc-bdVaJa krKcVG ant-input-group-wrapper"
                      >
                        <span
                          className="ant-input-wrapper ant-input-group"
                        >
                          <span
                            className="ant-input-group-addon"
                          >
                            var1
                          </span>
                          <input
                            className="ant-input"
                            disabled={false}
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            style={null}
                            type="text"
                            value=""
                          />
                        </span>
                      </span>
                    </Input>
                  </StyledComponent>
                </Styled(Input)>
                <Styled(Input)
                  addonBefore="var2"
                  key="var2"
                  onChange={[Function]}
                >
                  <StyledComponent
                    addonBefore="var2"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "Group": [Function],
                        "Password": [Function],
                        "Search": [Function],
                        "TextArea": [Function],
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bdVaJa",
                          "isStatic": true,
                          "lastClassName": "krKcVG",
                          "rules": Array [
                            "
  margin: 5px 5px 0 0;
  width: 200px;
",
                          ],
                        },
                        "displayName": "Styled(Input)",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bdVaJa",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onChange={[Function]}
                  >
                    <Input
                      addonBefore="var2"
                      className="sc-bdVaJa krKcVG"
                      disabled={false}
                      onChange={[Function]}
                      type="text"
                    >
                      <span
                        className="sc-bdVaJa krKcVG ant-input-group-wrapper"
                      >
                        <span
                          className="ant-input-wrapper ant-input-group"
                        >
                          <span
                            className="ant-input-group-addon"
                          >
                            var2
                          </span>
                          <input
                            className="ant-input"
                            disabled={false}
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            style={null}
                            type="text"
                            value=""
                          />
                        </span>
                      </span>
                    </Input>
                  </StyledComponent>
                </Styled(Input)>
                <Styled(Input)
                  addonBefore="var12"
                  key="var12"
                  onChange={[Function]}
                >
                  <StyledComponent
                    addonBefore="var12"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "Group": [Function],
                        "Password": [Function],
                        "Search": [Function],
                        "TextArea": [Function],
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bdVaJa",
                          "isStatic": true,
                          "lastClassName": "krKcVG",
                          "rules": Array [
                            "
  margin: 5px 5px 0 0;
  width: 200px;
",
                          ],
                        },
                        "displayName": "Styled(Input)",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bdVaJa",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onChange={[Function]}
                  >
                    <Input
                      addonBefore="var12"
                      className="sc-bdVaJa krKcVG"
                      disabled={false}
                      onChange={[Function]}
                      type="text"
                    >
                      <span
                        className="sc-bdVaJa krKcVG ant-input-group-wrapper"
                      >
                        <span
                          className="ant-input-wrapper ant-input-group"
                        >
                          <span
                            className="ant-input-group-addon"
                          >
                            var12
                          </span>
                          <input
                            className="ant-input"
                            disabled={false}
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            style={null}
                            type="text"
                            value=""
                          />
                        </span>
                      </span>
                    </Input>
                  </StyledComponent>
                </Styled(Input)>
              </div>
            </StyledComponent>
          </styled.div>
        </VariableInputs>
      </div>
    </StyledComponent>
  </styled.div>
</TemplatePreview>
`;
